File: smartrecyclerviewadapter/src/main/java/bobzien/com/smartrecyclerviewadapter/SmartRecyclerViewAdapter.java
package bobzien.com.smartrecyclerviewadapter;

import android.content.Context;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.RecyclerView.Adapter;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;

public class SmartRecyclerViewAdapter extends Adapter
{
    private HashMap<Class, ViewHolder> viewHolders = new HashMap<>();
    private HashMap<Integer, ViewHolder> itemViewTypeToViewHolders = new HashMap<>();
    private List<Object> items = new ArrayList<>();
    private List<TypeViewHolder> typeViewHolders = new ArrayList<>();
    private int selectedPosition;

    public SmartRecyclerViewAdapter(ViewHolder[] viewHolders) throws DuplicateItemViewTypeException
    {
        for (ViewHolder viewHolder : viewHolders)
        {
            addViewHolder(viewHolder);
        }
    }

    public void addViewHolders(ViewHolder... viewHolders)
    {
        for(ViewHolder viewHolder : viewHolders)
        {
            addViewHolder(viewHolder);
        }
    }

    public void addViewHolder(ViewHolder viewHolder)
    {
        if (itemViewTypeToViewHolders.containsKey(viewHolder.getGenericItemViewType()))
        {
            throw (new DuplicateItemViewTypeException(itemViewTypeToViewHolders.get(viewHolder.getGenericItemViewType()).getClass().getSimpleName()
                    + " has same ItemViewType as " + viewHolder.getClass().getSimpleName()
                    + ". Overwrite getGenericItemViewType() in one of the two classes and assign a new type. Consider using an android id."));
        }
        itemViewTypeToViewHolders.put(viewHolder.getGenericItemViewType(), viewHolder);

        if(viewHolder instanceof TypeViewHolder)
        {
            typeViewHolders.add((TypeViewHolder) viewHolder);
        }
        this.viewHolders.put(viewHolder.getHandledClass(), viewHolder);
    }

    public void setItems(List items)
    {
        this.items = new ArrayList<>(items);
        wrapItems();
        notifyDataSetChanged();
    }

    public void addItems(int position, Object ... newItems)
    {
        int insertionPosition = position;
        for(Object item : newItems)
        {
            items.add(position++, wrapItem(item));
        }
        notifyItemRangeInserted(insertionPosition, newItems.length);
    }

    public void addItems(int position, List newItems)
    {
        int insertionPosition = position;
        for(Object item : newItems)
        {
            items.add(position++, wrapItem(item));
        }
        wrapItems();
        notifyItemRangeInserted(insertionPosition, newItems.size());
    }

    public void removeItem(int position)
    {
        items.remove(position);
        notifyItemRemoved(position);
    }

    public void removeItemRange(int startPosition, int itemCount)
    {
        for(int i = 0; i < itemCount; i++)
        {
            items.remove(startPosition);
        }
        notifyItemRangeRemoved(startPosition, itemCount);
    }

    public Object getItem(int position)
    {
        if(position >= getItemCount() || position < 0)
        {
            return null;
        }

        Object item = items.get(position);
        if(item instanceof Wrapper)
        {
            return ((Wrapper)item).getItem();
        }
        return item;
    }

    private void wrapItems()
    {
        if(typeViewHolders.size() > 0)
        {
            int i = 0;
            for (Object item : items)
            {
                items.set(i, wrapItem(item));
                i++;
            }
        }
    }

    private Object wrapItem(Object item)
    {
        for(TypeViewHolder typeViewHolder : typeViewHolders)
        {
            if(typeViewHolder.internalCanHandle(item))
            {
                Wrapper wrappedObject = typeViewHolder.getWrappedObject(item);
                if(wrappedObject == null)
                {
                    throw new WrapperNullException("The wrapper returned by " + typeViewHolder.getClass().getSimpleName() + " was null."
                    + " Please implement the method getWrappedObject and let it return an inner class which implements Wrapper."
                    + " The method getItem() should return the item that was passed into the getWrappedObject() method as a parameter.");
                }
                return wrappedObject;
            }
        }
        return item;
    }

    @Override
    public ViewHolder onCreateViewHolder(ViewGroup parent, int viewType)
    {
        ViewHolder viewHolder = itemViewTypeToViewHolders.get(viewType);

        ViewHolder instance = viewHolder.getInstance(parent);
        instance.setAdapter(this);
        return instance;
    }

    @Override
    public void onBindViewHolder(RecyclerView.ViewHolder holder, int position)
    {
        Object item = items.get(position);
        if(holder instanceof TypeViewHolder)
        {
            item = ((Wrapper)item).getItem();
        }

        ((ViewHolder)holder).bindViewHolder(item, position == selectedPosition);
    }

    @Override
    public int getItemCount()
    {
        return items.size();
    }

    @Override
    public int getItemViewType(int position)
    {
        ViewHolder viewHolder = getViewHolder(position);

        return viewHolder.getGenericItemViewType();
    }

    private ViewHolder getViewHolder(int position)
    {
        Object item = items.get(position);

        Class<?> itemClass = item.getClass();

        ViewHolder viewHolder = viewHolders.get(itemClass);

        if(viewHolder == null)
        {
            viewHolder = tryToFindSuperClass(itemClass);
            if (viewHolder != null)
            {
                return viewHolder;
            }
            throw(new MissingViewHolderException("There is no Viewholder for " + itemClass.getSimpleName()
                    + " or one of its superClasses."));
        }
        return viewHolder;
    }

    private ViewHolder tryToFindSuperClass(Class<?> itemClass)
    {
        ViewHolder viewHolder;
        Class<?> superclass = itemClass.getSuperclass();
        while(superclass != null)
        {
            viewHolder = viewHolders.get(superclass);
            if(viewHolder != null)
            {
                viewHolders.put(superclass, viewHolder);
                return viewHolder;
            }
            superclass = superclass.getSuperclass();
        }
        return null;
    }

    public int getSelectedPosition()
    {
        return selectedPosition;
    }

    public void setSelectedPosition(int selectedPosition)
    {
        int oldSelectedPosition = this.selectedPosition;
        this.selectedPosition = selectedPosition;
        this.selectedPosition = selectedPosition;

        notifyItemChanged(oldSelectedPosition);
        notifyItemChanged(selectedPosition);
    }

    public static abstract class ViewHolder<T> extends RecyclerView.ViewHolder
    {
        private Context context;
        private Class clazz;
        private SmartRecyclerViewAdapter adapter;

        /**
         * This constructor is just used to create a factory that can produce the
         * ViewHolder that contains the actual itemView using the getInstance(View parent) method.
         * @param context
         * @param clazz
         */
        public ViewHolder(@NonNull Context context, @NonNull Class clazz, @Nullable View itemView)
        {
            super(itemView == null ? new View(context) : itemView);
            this.context = context;
            this.clazz = clazz;
        }

        private ViewHolder getInstance(@NonNull ViewGroup parent)
        {
            View view = LayoutInflater.from(getContext()).inflate(getLayoutResourceId(), parent, false);
            return getInstance(view);
        }

        /**
         * This has to create a new instance which was created using a constructor which calls super(itemView).
         * @param itemView
         * @return the new ViewHolder
         */
        public abstract ViewHolder getInstance(@NonNull View itemView);

        /**
         * Gets the LayoutResourceId of the layout this ViewHolder uses
         * @return
         */
        public abstract int getLayoutResourceId();

        /**
         * Override this if itemViewTypes are colliding. Consider using an android Id.
         * @return itemViewType
         */
        public int getGenericItemViewType()
        {
            return getLayoutResourceId();
        }

        /**
         * Bind the ViewHolder and return it.
         * @param item
         * @param selected
         * @return ViewHolder
         */
        @NonNull
        public abstract void bindViewHolder(@NonNull T item, boolean selected);

        protected Class getHandledClass()
        {
            return clazz;
        }

        public Context getContext()
        {
            return context;
        }

        public SmartRecyclerViewAdapter getAdapter()
        {
            return adapter;
        }
        public void setAdapter(SmartRecyclerViewAdapter adapter)
        {
            this.adapter = adapter;
        }
    }

    public static abstract class TypeViewHolder<T> extends ViewHolder<T>
    {
        public TypeViewHolder(@NonNull Context context, @NonNull Class clazz, @Nullable View itemView)
        {
            super(context, clazz, itemView);
        }

        private boolean internalCanHandle(@NonNull Object item)
        {
            if(item.getClass().equals(super.getHandledClass()))
            {
                return canHandle((T)item);
            }
            return false;
        }

        protected abstract boolean canHandle(@NonNull T item);

        @NonNull
        protected abstract Wrapper getWrappedObject(@Nullable final T item);

        @Override
        protected final Class getHandledClass()
        {
            return getWrappedObject(null).getClass();
        }
    }

    public interface Wrapper
    {
        Object getItem();
    }

    public static class DuplicateItemViewTypeException extends RuntimeException
    {
        public DuplicateItemViewTypeException(String detailMessage)
        {
            super(detailMessage);
        }
    }

    public static class MissingViewHolderException extends RuntimeException
    {
        public MissingViewHolderException(String detailMessage)
        {
            super(detailMessage);
        }
    }

    public static class WrapperNullException extends RuntimeException
    {
        public WrapperNullException(String detailMessage)
        {
            super(detailMessage);
        }
    }
}

File: smartrecyclerviewadapter/src/main/kotlin/bobzien/com/smartrecyclerviewadapter/SmartRecyclerViewAdapter.kt
package bobzien.com.smartrecyclerviewadapter

import android.content.Context
import android.support.v7.widget.RecyclerView
import android.support.v7.widget.RecyclerView.Adapter
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup

import java.util.ArrayList
import java.util.HashMap

class SmartRecyclerViewAdapter @Throws(SmartRecyclerViewAdapter.DuplicateItemViewTypeException::class)
constructor(viewHolders: Array<SmartRecyclerViewAdapter.ViewHolder<Any>>) : Adapter<SmartRecyclerViewAdapter.ViewHolder<Any>>() {
    private val viewHolders = HashMap<Class<Any>, ViewHolder<Any>>()
    private val itemViewTypeToViewHolders = HashMap<Int, ViewHolder<Any>>()
    private var items: MutableList<Any> = ArrayList()
    private val typeViewHolders = ArrayList<TypeViewHolder<Any>>()
    var selectedPosition: Int = 0
        set(selectedPosition: Int) {
            val oldSelectedPosition = this.selectedPosition
            field = selectedPosition

            notifyItemChanged(oldSelectedPosition)
            notifyItemChanged(selectedPosition)
        }

    init {
        for (viewHolder in viewHolders) {
            addViewHolder(viewHolder)
        }
    }

    fun addViewHolders(vararg viewHolders: ViewHolder<Any>) {
        for (viewHolder in viewHolders) {
            addViewHolder(viewHolder)
        }
    }

    fun addViewHolder(viewHolder: ViewHolder<Any>) {
        if (itemViewTypeToViewHolders.containsKey(viewHolder.genericItemViewType)) {
            throw (DuplicateItemViewTypeException((itemViewTypeToViewHolders.get(viewHolder.genericItemViewType) as Any).javaClass.simpleName
                    + " has same ItemViewType as " + viewHolder.javaClass.simpleName
                    + ". Overwrite getGenericItemViewType() in one of the two classes and assign a new type. Consider using an android id."))
        }
        itemViewTypeToViewHolders.put(viewHolder.genericItemViewType, viewHolder)

        if (viewHolder is TypeViewHolder<Any>) {
            typeViewHolders.add(viewHolder)
        }
        this.viewHolders.put(viewHolder.getHandledClass() as Class<Any>, viewHolder)
    }

    fun setItems(items: List<Any>) {
        this.items = ArrayList(items)
        wrapItems()
        notifyDataSetChanged()
    }

    fun addItems(position: Int, vararg newItems: Any) {
        var index = position
        val insertionPosition = position
        for (item in newItems) {
            items.add(index++, wrapItem(item))
        }
        notifyItemRangeInserted(insertionPosition, newItems.size())
    }

    fun addItems(position: Int, newItems: List<Any>) {
        var index = position
        val insertionPosition = position
        for (item in newItems) {
            items.add(index++, wrapItem(item))
        }
        wrapItems()
        notifyItemRangeInserted(insertionPosition, newItems.size())
    }

    fun removeItem(position: Int) {
        items.remove(position)
        notifyItemRemoved(position)
    }

    fun removeItemRange(startPosition: Int, itemCount: Int) {
        for (i in 0..itemCount - 1) {
            items.remove(startPosition)
        }
        notifyItemRangeRemoved(startPosition, itemCount)
    }

    fun getItem(position: Int): Any? {
        if (position >= itemCount || position < 0) {
            return null
        }

        val item = items.get(position)
        if (item is Wrapper) {
            return item.item
        }
        return item
    }

    private fun wrapItems() {
        if (typeViewHolders.size() > 0) {
            var i = 0
            for (item in items) {
                items.set(i, wrapItem(item))
                i++
            }
        }
    }

    private fun wrapItem(item: Any): Any {
        for (typeViewHolder in typeViewHolders) {
            if (typeViewHolder.internalCanHandle(item)) {
                val wrappedObject = typeViewHolder.getWrappedObject(item)
                return wrappedObject
            }
        }
        return item
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder<Any> {
        val viewHolder = itemViewTypeToViewHolders.get(viewType)

        val instance = viewHolder!!.getInstance(parent)
        instance.adapter = this
        return instance
    }

    override fun onBindViewHolder(holder: ViewHolder<Any>?, position: Int) {
        var item = items.get(position)
        if (holder is TypeViewHolder<*>) {
            item = (item as Wrapper).item
        }

        (holder as ViewHolder<Any>).bindViewHolder(item, position == selectedPosition)
    }

    override fun getItemCount(): Int {
        return items.size()
    }

    override fun getItemViewType(position: Int): Int {
        val viewHolder = getViewHolder(position)

        return viewHolder.genericItemViewType
    }

    private fun getViewHolder(position: Int): ViewHolder<Any> {
        val item = items.get(position)

        val itemClass = item.javaClass

        var viewHolder: ViewHolder<Any>? = viewHolders.get(itemClass)

        if (viewHolder == null) {
            viewHolder = tryToFindSuperClass(itemClass)
            if (viewHolder != null) {
                return viewHolder
            }
            throw (MissingViewHolderException("There is no Viewholder for " + itemClass.simpleName + " or one of its superClasses."))
        }
        return viewHolder
    }

    private fun tryToFindSuperClass(itemClass: Class<*>): ViewHolder<Any>? {
        var viewHolder: ViewHolder<Any>?
        var superclass: Class<*>? = itemClass.superclass
        while (superclass != null) {
            viewHolder = viewHolders.get(superclass)
            if (viewHolder != null) {
                viewHolders.put(superclass as Class<Any>, viewHolder)
                return viewHolder
            }
            superclass = superclass.superclass
        }
        return null
    }

    abstract class ViewHolder<T>
    /**
     * This constructor is just used to create a factory that can produce the
     * ViewHolder that contains the actual itemView using the getInstance(View parent) method.
     * @param context
     * *
     * @param clazz
     */
    (val context: Context, handledClass: Class<*>, itemView: View?) : RecyclerView.ViewHolder(itemView ?: View(context)) {

        var adapter: SmartRecyclerViewAdapter? = null

        private var _handledClass: Class<*>
        open fun getHandledClass(): Class<*>{
            return _handledClass
        }

        init {
            _handledClass = handledClass
        }

        fun getInstance(parent: ViewGroup): ViewHolder<Any> {
            val view = LayoutInflater.from(context).inflate(layoutResourceId, parent, false)
            return getInstance(view)
        }

        /**
         * This has to create a new instance which was created using a constructor which calls super(itemView).
         * @param itemView
         * *
         * @return the new ViewHolder
         */
        abstract fun getInstance(itemView: View): ViewHolder<Any>

        /**
         * Gets the LayoutResourceId of the layout this ViewHolder uses
         * @return
         */
        abstract val layoutResourceId: Int

        /**
         * Override this if itemViewTypes are colliding. Consider using an android Id.
         * @return itemViewType
         */
        val genericItemViewType: Int
            get() = layoutResourceId

        /**
         * Bind the ViewHolder and return it.
         * @param item
         * *
         * @param selected
         * *
         * @return ViewHolder
         */
        abstract fun bindViewHolder(item: T, selected: Boolean)
    }

    abstract class TypeViewHolder<T>(context: Context, clazz: Class<Any>, itemView: View?) : ViewHolder<T>(context, clazz, itemView) {

        private lateinit var _handledClass: Class<*>

        init{
            _handledClass = getWrappedObject(null).javaClass
        }

        fun internalCanHandle(item: Any): Boolean {
            if (item.javaClass == super.getHandledClass()) {
                return canHandle(item as T)
            }
            return false
        }

        protected abstract fun canHandle(item: T): Boolean

        abstract fun getWrappedObject(item: T?): Wrapper

        override fun getHandledClass(): Class<*>{
            return _handledClass
        }
    }

    interface Wrapper {
        val item: Any
    }

    class DuplicateItemViewTypeException(detailMessage: String) : RuntimeException(detailMessage)

    class MissingViewHolderException(detailMessage: String) : RuntimeException(detailMessage)
}
